### PR TITLE
50 unit test for bmd boundary cases

### PR DESCRIPTION
**Pull request by gracezhihuizhao**
Added unit tests to cover cases listed in #50.

### CODE REVIEW

File: tests/testthat/test-tcplhit2_core.R [hunk 1]
```diff
@@ -13,3 +13,133 @@ test_that("tcplhit2 works", {
...
+  expect_equal(output_after$bmdl, 0.2826, tolerance = 1e-3)
+  expect_equal(output_after_ten$bmdl, 0.1048, tolerance = 1e-3)
+
+  # checks for bmd confident upper bound
+  expect_equal(output_before$bmdu, 0.815, tolerance = 1e-3)
+  expect_equal(output_after$bmdu, 0.9928, tolerance = 1e-3)
+  expect_equal(output_after_ten$bmdu, 0.815, tolerance = 1e-3)
+})
+
+test_that("tcplhit2 BMD boundary censor checks", {
```
Comment (sedavid01) [inline tests/testthat/test-tcplhit2_core.R:106]: Suggest adding similar test but add the lower and upper boundary that are 0.1 and 10.

Comment (sedavid01) [reply tests/testthat/test-tcplhit2_core.R:106]: This tests I would suggest updating to "tcplhit2 BMD boundary check - no censoring needed".  Then the other one (suggested above) "tcplhit2 BMD boundary check - censoring required".  Or something to that effect.

Let me know if you have further questions on this.

**Review by sedavid01 — CHANGES_REQUESTED** [review]
Overall this is looking good thus far.  There is a minor suggestion for updating the tests.

File: tests/testthat/test-tcplhit2_core.R
Comment (sedavid01) [inline tests/testthat/test-tcplhit2_core.R:1]: Look into whether there is a way to streamline some of the repeated code/data pulling/management between these new tests, such that the same code used for set-up across these tests can be called and reused focusing just on the different argument calls and expectation.

File: tests/testthat/test-tcplhit2_core.R [hunk 2]
```diff
@@ -13,3 +13,158 @@ test_that("tcplhit2 works", {
   expect_equal(output$tp, 0.749, tolerance = 1e-2)
   expect_equal(output$ga, 9.59, tolerance = 1e-2)
 })
+
+# Preparation for test on BMD boundary arguments
+data("mc3")
+
```
[the remark this answers is not in the document]
Comment (sedavid01) [reply tests/testthat/test-tcplhit2_core.R:19]: mc3_UT <- copy(mc3) # if mc3 is a data.table object class.

**Review by sedavid01 — CHANGES_REQUESTED** [review]
Slight update to better capture the tests that are needed and I think capture the behavior we want to ensure.  Please let me know if you have questions.

File: tests/testthat/test-tcplhit2_core.R [hunk 3]
```diff
@@ -13,3 +13,150 @@ test_that("tcplhit2 works", {
...
 })
+
+test_that("tcplhit2 BMD boundary check", {
+  # Preparation for tests on BMD boundary arguments
+  # Simulate some data
+  X <- rep(seq(0.001,3,length.out = 10),each = 5)
+
+  set.seed(918)
+  Case_1 <- hillfn(ps = c(4,0.00105,1.08,0.1),x = X) + rt(n = length(X),df = 4)
+  Case_3 <- pow(ps = c(10,5,0.1),x = X) + rt(n = length(X),df = 4)
```
Comment (sedavid01) [inline tests/testthat/test-tcplhit2_core.R:24]: Do you need to add a set.seed() before this line setting Case 3?

File: tests/testthat/test-tcplhit2_core.R [hunk 4]
```diff
@@ -13,3 +13,150 @@ test_that("tcplhit2 works", {
...
+  df <- matrix(nrow = 5, ncol = 14)
+  colnames(df) <- c("onesd", "cutoff", "bmd", "bmdu", "bmdl",
+                    "combo1", "combo1u", "combo1l",
+                    "combo2", "combo2u", "combo2l",
+                    "combo3", "combo3u", "combo3l")
+
+  # fit each data case and hit-call with different argument combinations
+  for (i in 1:5) {
+    temp <- Y[i, 1:10]
+    bmad <- mad(temp)
```
Comment (sedavid01) [inline tests/testthat/test-tcplhit2_core.R:52]: BMAD should be mad(Y[,1:10]), that is the median absolute deviation of replicates in the two lowest concentrations across all responses - NOT the median absolute deviation of replicates in the two lowest concentrations for a single response.

The bmad and the cutoff can be calculated prior to the for loop.  The cutoff assignment to the 'df' matrix can be done prior to the for loop as well.

File: tests/testthat/test-tcplhit2_core.R [hunk 5]
```diff
@@ -13,3 +13,150 @@ test_that("tcplhit2 works", {
...
+    params <- tcplfit2_core(X, Y[i, ], cutoff,
+                            force.fit = T, bidirectional = F)
+    # No thresholds specified, no censoring
+    output <- tcplhit2_core(params, X, Y[i,], cutoff, onesd,
+                            bmed=0, conthits=T, aicc=F)
+
+    df[i, "bmd"] <- output$bmd
+    df[i, "bmdu"] <- output$bmdu
+    df[i, "bmdl"] <- output$bmdl
+
```
Comment (sedavid01) [inline tests/testthat/test-tcplhit2_core.R:68]: Suggest putting a brief comment for the choice of bmd boundary multiples.  For example,
"# For these tests we use a lower and upper BMD threshold multiple of 0.7 and 10, respectively, to demonstrate use cases work for the simulated concentration-response data.  Threshold multiples are not necessarily meant to align with the typical suggested values of 0.1 and 10 for lower and upper threshold multiples, respectively."

**Review by sedavid01 — CHANGES_REQUESTED** [review]
Minor updates requested prior to approval.

File: tests/testthat/test-tcplhit2_core.R (hunk 4 above)
Comment (gracezhihuizhao) [reply tests/testthat/test-tcplhit2_core.R:52]: @sedavid01 Y is a 5x50 matrix, each row is one data case (10 doses, 5 replicates each). I thought each data case should be independent and calculate bmed and cutoff separately. `Y[i, 1:10]` are replicates in the two lowest concentrations for the i-th data case. Let me know if I misunderstood.

Comment (sedavid01) [reply tests/testthat/test-tcplhit2_core.R:52]: Each row in your Y is an individual response (i.e. data case or response case).

- onesd - calculated on the individual response (single row :: single response)
- bmad & cutoff - across a group (or all) of responses in the dataset (all rows :: all responses)

File: tests/testthat/test-tcplhit2_core.R
Comment (sedavid01) [inline tests/testthat/test-tcplhit2_core.R:1]: @gracezhihuizhao, can you revisit this PR to review and ensure all comments are addressed.  I think there is just one outstanding comment, but a wholistic review would be appreciated so we can close this ticket/branch out.

Comment (gracezhihuizhao) [reply tests/testthat/test-tcplhit2_core.R:1]: @sedavid01 I've reviewed the pull request. All comments have been addressed.

File: tests/testthat/test-tcplhit2_core.R (hunk 5 above)
Comment (gracezhihuizhao) [reply tests/testthat/test-tcplhit2_core.R:78]: Comment added in the latest commit.

Comment (sedavid01) [reply tests/testthat/test-tcplhit2_core.R:78]: See it now, thank you for confirming @gracezhihuizhao.

**Review by sedavid01 — APPROVED** [review]
All comments have been addressed.  Approving this PR and ready for merging.

**Comment by sedavid01** [comment]
@brown-jason, changes are merged to 'dev' please verify when you get the chance and if any branch clean-up is needed please either do so or provide guidance/instructions. Thanks, -Sarah
